File: librarymanager/librarymanager-be/src/main/java/ru/spring5/elena/librarymanager/services/BookService.java
package ru.spring5.elena.librarymanager.services;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import ru.spring5.elena.librarymanager.api.dto.Book;
import ru.spring5.elena.librarymanager.api.dto.BookUsing;
import ru.spring5.elena.librarymanager.domain.entity.BookEntity;
import ru.spring5.elena.librarymanager.domain.entity.EditionEntity;
import ru.spring5.elena.librarymanager.domain.repository.BookRepository;
import ru.spring5.elena.librarymanager.domain.repository.BookUsingRepository;
import ru.spring5.elena.librarymanager.domain.repository.EditionRepository;
import ru.spring5.elena.librarymanager.mapper.BookMapper;
import ru.spring5.elena.librarymanager.mapper.BookUsingMapper;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;


@RequiredArgsConstructor
@Service
@Slf4j
public class BookService {

    private final BookRepository bookRepository;
    private final BookUsingRepository bookUsingRepository;
    private final EditionRepository editionRepository;

    @Transactional(readOnly = true)
    public List<Book> findAll() {
        return bookRepository.findAll()
                .stream()
                .map(BookMapper::map)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public Page<Book> findAllByPage(Pageable pageable) {
        return bookRepository.findAll(pageable)
                .map(BookMapper::map);
    }

    @Transactional(readOnly = true)
    public Book findBookById(Long id) {
        return bookRepository.findById(id)
                .map(BookMapper::map)
                .orElse(null);
    }

    public Book save(Long bookId, Book dto) {
    	EditionEntity edition = editionRepository.findById(dto.getEditionId()).get();
    	return bookRepository.findById(bookId)
                .map(bookEntity -> BookMapper.map(dto, bookEntity, edition))
                .map(bookRepository::saveAndFlush)
                .map(BookMapper::map)
                .orElse(null);
    }

    public Book create(Book dto) {
    	EditionEntity edition = editionRepository.findById(dto.getEditionId()).get();
    	return Optional.of(dto)
                .map(book -> BookMapper.map(book,edition))
                .map(bookRepository::saveAndFlush)
                .map(BookMapper::map)
                .get();
    }

    @Transactional(readOnly = true)
	public Page<BookUsing> findBookUsingsByBookId(Long bookId, Pageable pageable) {
		return bookRepository.findById(bookId)
                .map(entity -> bookUsingRepository.findAllByBook(entity, pageable))
                .orElse(Page.empty())
                .map(BookUsingMapper:: map);
	}

    @Transactional
	public List<BookUsing> updateBookUsings(Long bookId, List<Long> bookUsingIds) {
		return bookRepository.findById(bookId)
		.map(bookEntity-> {
			bookEntity.getBookUsings().clear();
			bookEntity.getBookUsings().addAll(
					bookUsingIds.stream()
						.map(bookUsingRepository::getOne)
						.collect(Collectors.toList())
						);
			return bookEntity;
		 })
		.map(bookRepository::saveAndFlush)
		.map(BookEntity :: getBookUsings)
		.orElse(Collections.emptySet())
		.stream()
		.map(BookUsingMapper :: map)
		.collect(Collectors.toList());
	}

}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/services/BookService.kt
package ru.elenacherev.librarymanager.services

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.elenacherev.librarymanager.api.dto.Book
import ru.elenacherev.librarymanager.api.dto.BookUsing
import ru.elenacherev.librarymanager.domain.entity.BookEntity
import ru.elenacherev.librarymanager.domain.entity.BookUsingEntity
import ru.elenacherev.librarymanager.domain.entity.EditionEntity
import ru.elenacherev.librarymanager.domain.repository.BookRepository
import ru.elenacherev.librarymanager.domain.repository.BookUsingRepository
import ru.elenacherev.librarymanager.domain.repository.EditionRepository
import ru.elenacherev.librarymanager.mapper.map
import java.util.*

@Service
class BookService(
    val bookRepository: BookRepository,
    val bookUsingRepository: BookUsingRepository,
    val editionRepository: EditionRepository
) {
    @Transactional(readOnly = true)
    fun findAll(): List<Book> {
        return bookRepository
            .findAll()
            .map(BookEntity::map)
    }

    @Transactional(readOnly = true)
    fun findAllByPage(pageable: Pageable): Page<Book> {
        return bookRepository
            .findAll(pageable)
            .map(BookEntity::map)
    }

    @Transactional(readOnly = true)
    fun findBookById(id: UUID): Book {
        return bookRepository
            .findById(id)
            .map(BookEntity::map)
            .orElse(null)
    }

    fun save(bookId: UUID, dto: Book): Book {
        val edition: EditionEntity = editionRepository.findById(dto.editionId).get()
        return bookRepository
            .findById(bookId)
            .map { dto.map(edition) }
            .map { book: BookEntity -> bookRepository.saveAndFlush(book) }
            .map(BookEntity::map)
            .orElse(null)
    }

    fun create(dto: Book): Book {
        val edition: EditionEntity = editionRepository.findById(dto.editionId).get()
        return Optional.of(dto)
            .map { dto.map(edition) }
            .map { book: BookEntity -> bookRepository.saveAndFlush(book) }
            .map(BookEntity::map)
            .get()
    }

    @Transactional(readOnly = true)
    fun findBookUsingsByBookId(bookId: UUID, pageable: Pageable): Page<BookUsing> {
        return bookRepository.findById(bookId)
            .map { entity -> bookUsingRepository.findAllByBook(entity, pageable) }
            .orElse(Page.empty())
            .map(BookUsingEntity::map)
    }

    @Transactional
    fun updateBookUsings(bookId: UUID, bookUsingIds: List<UUID>): List<BookUsing> {
        return bookRepository.findById(bookId)
            .map { bookEntity ->
                bookEntity.bookUsings.clear()
                bookEntity.bookUsings += bookUsingIds.map(bookUsingRepository::getOne)
                return@map bookEntity
            }
            .map { book: BookEntity -> bookRepository.saveAndFlush(book) }
            .map(BookEntity::bookUsings)
            .orElse(Collections.emptySet())
            .map(BookUsingEntity::map)
    }
}